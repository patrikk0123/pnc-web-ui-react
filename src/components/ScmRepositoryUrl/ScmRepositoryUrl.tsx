import { Button } from '@patternfly/react-core';
import { ExternalLinkAltIcon } from '@patternfly/react-icons';

import { SCMRepository } from 'pnc-api-types-ts';

import { CopyToClipboard } from 'components/CopyToClipboard/CopyToClipboard';
import { TooltipWrapper } from 'components/TooltipWrapper/TooltipWrapper';

import { uiLogger } from 'services/uiLogger';

import { IParsedUrl, parseScmRepositoryUrl } from 'utils/urlParseHelper';

interface IUrlButtonProps {
  parsedUrl: IParsedUrl;
  isInline?: boolean;
}

/**
 * The Url button that redirect to specific page.
 *
 * @param parsedUrl - the object contains url and name
 * @param isInline - whether to use inline style with external link action
 */
const UrlButton = ({ parsedUrl, isInline }: IUrlButtonProps) => (
  <TooltipWrapper tooltip={'View in ' + parsedUrl.name}>
    <Button
      component="a"
      href={parsedUrl.webUrl}
      target="_blank"
      rel="noopener noreferrer"
      variant={isInline ? 'plain' : 'tertiary'}
      hasNoPadding={isInline}
      icon={<ExternalLinkAltIcon />}
    >
      {!isInline && parsedUrl.name}
    </Button>
  </TooltipWrapper>
);

interface IScmRepositoryUrlProps {
  internalScmRepository?: SCMRepository;
  externalScmRepository?: SCMRepository;
  isInline?: boolean;
}

/**
 * Represents the internal/external URL for the SCM Repository.
 *
 * @param internalScmRepository - the SCM Repository to parse its internal url
 * @param externalScmRepository - the SCM Repository to parse its external url
 * @param isInline - whether to use inline style with external link action
 */
export const ScmRepositoryUrl = ({ internalScmRepository, externalScmRepository, isInline }: IScmRepositoryUrlProps) => {
  let parsedUrl;
  if (internalScmRepository) {
    parsedUrl = parseScmRepositoryUrl({ url: internalScmRepository.internalUrl });
  } else if (externalScmRepository) {
    parsedUrl = externalScmRepository.externalUrl ? parseScmRepositoryUrl({ url: externalScmRepository.externalUrl }) : null;
  } else {
    uiLogger.error('internalScmRepository or externalScmRepository has to be defined');
  }

  return parsedUrl ? (
    <CopyToClipboard isInline={isInline} suffixComponent={<UrlButton isInline={isInline} parsedUrl={parsedUrl} />}>
      {parsedUrl.scmRepositoryUrl}
    </CopyToClipboard>
  ) : null;
};
