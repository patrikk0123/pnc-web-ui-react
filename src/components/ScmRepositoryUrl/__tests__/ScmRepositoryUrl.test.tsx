import { render } from '@testing-library/react';
import { vi } from 'vitest';

import { ScmRepositoryUrl } from 'components/ScmRepositoryUrl/ScmRepositoryUrl';

vi.mock('services/uiLogger');

describe('display ScmRepositoryUrl component', () => {
  test('renders ScmRepositoryLink', () => {
    render(
      <ScmRepositoryUrl
        internalScmRepository={{
          id: '103',
          internalUrl: 'git+ssh://test.env.com/testRepo/testUrlClipboardCopyGerrit.git',
          externalUrl: 'https://github.com/testRepo/empty.git',
          preBuildSyncEnabled: true,
        }}
      />
    );
  });

  test('compare snapshot with previous record', () => {
    const tree = render(
      <ScmRepositoryUrl
        internalScmRepository={{
          id: '103',
          internalUrl: 'git+ssh://test.env.com/testRepo/testUrlClipboardCopyGerrit.git',
          externalUrl: 'https://github.com/testRepo/empty.git',
          preBuildSyncEnabled: true,
        }}
      />
    );
    expect(tree.container).toMatchSnapshot();
  });
});
