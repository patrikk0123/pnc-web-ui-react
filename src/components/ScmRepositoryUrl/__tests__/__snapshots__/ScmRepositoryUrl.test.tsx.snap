// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display ScmRepositoryUrl component compare snapshot with previous record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="pf-v5-l-split pf-m-gutter"
      >
        <div
          class="pf-v5-l-split__item pf-m-fill"
        >
          <div
            class="pf-v5-c-clipboard-copy no-bg"
            data-ouia-component-type="PF5/ClipboardCopy"
            data-ouia-safe="true"
          >
            <div
              class="pf-v5-c-clipboard-copy__group"
            >
              <span
                class="pf-v5-c-form-control pf-m-readonly"
              >
                <input
                  aria-invalid="false"
                  aria-label="Copyable input"
                  data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                  data-ouia-component-type="PF5/TextInput"
                  data-ouia-safe="true"
                  id="text-input-1"
                  readonly=""
                  type="text"
                  value="git+ssh://test.env.com/testRepo/testUrlClipboardCopyGerrit.git"
                />
              </span>
              <button
                aria-disabled="false"
                aria-label="Copy"
                aria-labelledby="copy-button-1 text-input-1"
                class="pf-v5-c-button pf-m-control"
                data-ouia-component-id="OUIA-Generated-Button-control-2"
                data-ouia-component-type="PF5/Button"
                data-ouia-safe="true"
                id="copy-button-1"
                type="button"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 448 512"
                  width="1em"
                >
                  <path
                    d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="pf-v5-l-split__item"
        >
          <div
            style="display: contents;"
          >
             
            <a
              aria-disabled="false"
              class="pf-v5-c-button pf-m-tertiary"
              data-ouia-component-id="OUIA-Generated-Button-tertiary-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              href="https://test.env.com/testRepo/testUrlClipboardCopyGerrit.git"
              rel="noopener noreferrer"
              target="_blank"
            >
              <span
                class="pf-v5-c-button__icon pf-m-start"
              >
                <svg
                  aria-hidden="true"
                  class="pf-v5-svg"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                  />
                </svg>
              </span>
              test.env.com
            </a>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="pf-v5-l-split pf-m-gutter"
    >
      <div
        class="pf-v5-l-split__item pf-m-fill"
      >
        <div
          class="pf-v5-c-clipboard-copy no-bg"
          data-ouia-component-type="PF5/ClipboardCopy"
          data-ouia-safe="true"
        >
          <div
            class="pf-v5-c-clipboard-copy__group"
          >
            <span
              class="pf-v5-c-form-control pf-m-readonly"
            >
              <input
                aria-invalid="false"
                aria-label="Copyable input"
                data-ouia-component-id="OUIA-Generated-TextInputBase-2"
                data-ouia-component-type="PF5/TextInput"
                data-ouia-safe="true"
                id="text-input-1"
                readonly=""
                type="text"
                value="git+ssh://test.env.com/testRepo/testUrlClipboardCopyGerrit.git"
              />
            </span>
            <button
              aria-disabled="false"
              aria-label="Copy"
              aria-labelledby="copy-button-1 text-input-1"
              class="pf-v5-c-button pf-m-control"
              data-ouia-component-id="OUIA-Generated-Button-control-2"
              data-ouia-component-type="PF5/Button"
              data-ouia-safe="true"
              id="copy-button-1"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 448 512"
                width="1em"
              >
                <path
                  d="M320 448v40c0 13.255-10.745 24-24 24H24c-13.255 0-24-10.745-24-24V120c0-13.255 10.745-24 24-24h72v296c0 30.879 25.121 56 56 56h168zm0-344V0H152c-13.255 0-24 10.745-24 24v368c0 13.255 10.745 24 24 24h272c13.255 0 24-10.745 24-24V128H344c-13.2 0-24-10.8-24-24zm120.971-31.029L375.029 7.029A24 24 0 0 0 358.059 0H352v96h96v-6.059a24 24 0 0 0-7.029-16.97z"
                />
              </svg>
            </button>
          </div>
        </div>
      </div>
      <div
        class="pf-v5-l-split__item"
      >
        <div
          style="display: contents;"
        >
           
          <a
            aria-disabled="false"
            class="pf-v5-c-button pf-m-tertiary"
            data-ouia-component-id="OUIA-Generated-Button-tertiary-2"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
            href="https://test.env.com/testRepo/testUrlClipboardCopyGerrit.git"
            rel="noopener noreferrer"
            target="_blank"
          >
            <span
              class="pf-v5-c-button__icon pf-m-start"
            >
              <svg
                aria-hidden="true"
                class="pf-v5-svg"
                fill="currentColor"
                height="1em"
                role="img"
                viewBox="0 0 512 512"
                width="1em"
              >
                <path
                  d="M432,320H400a16,16,0,0,0-16,16V448H64V128H208a16,16,0,0,0,16-16V80a16,16,0,0,0-16-16H48A48,48,0,0,0,0,112V464a48,48,0,0,0,48,48H400a48,48,0,0,0,48-48V336A16,16,0,0,0,432,320ZM488,0h-128c-21.37,0-32.05,25.91-17,41l35.73,35.73L135,320.37a24,24,0,0,0,0,34L157.67,377a24,24,0,0,0,34,0L435.28,133.32,471,169c15,15,41,4.5,41-17V24A24,24,0,0,0,488,0Z"
                />
              </svg>
            </span>
            test.env.com
          </a>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
